phoenix-core/src/main/java/org/apache/phoenix/expression/util/regex/JONIRegexWrapper.java

```diff
@@ -0,0 +1,181 @@
+                case ZERO:
+                    switch (ch) {
+                    case '\\':
+                        return ONE_BACKSLASH;
+                    default:
+                        sb.append(ch);
+                        return ZERO;
+                    }
+                case ONE_BACKSLASH:
+                    switch (ch) {
+                    case 'Q':
+                        return IN_QUOTE;
+                    default:
+                        sb.append('\\');
+                        return ZERO.next(sb, ch);
+                    }
+                case IN_QUOTE:
+                    // add backslashes for .^$*+?()[{\|
+                    switch (ch) {
+                    case '.':
+                    case '^':
+                    case '$':
+                    case '*':
+                    case '+':
+                    case '?':
+                    case '(':
+                    case ')':
+                    case '[':
+                    case '{':
+                    case '|':
+                        sb.append('\\').append(ch);
+                        return IN_QUOTE;
+                    case '\\':
+                        return IN_QUOTE_ONE_BACKSLASH;
+                    default:
+                        sb.append(ch);
+                        return IN_QUOTE;
+                    }
+                case IN_QUOTE_ONE_BACKSLASH:
+                    switch (ch) {
+                    case 'E':
+                        return ZERO;
+                    default:
+                        sb.append('\\');
+                        return IN_QUOTE.next(sb, ch);
+                    }
+                case STOP:
+                default:
+                    throw new IllegalArgumentException();
+                }
+            }
+
+            ReplaceQuoteMachine EOF(StringBuilder sb) {
+                switch (this) {
+                case ONE_BACKSLASH:
+                case IN_QUOTE_ONE_BACKSLASH:
+                    sb.append('\\');
+                case ZERO:
+                case IN_QUOTE:
+                    return STOP;
+                case STOP:
+                default:
+                    throw new IllegalArgumentException();
+                }
+            }
+        }
+
+        private String replacePatternQuote(String patternString) {
+            StringBuilder sb = new StringBuilder();
+            ReplaceQuoteMachine cur = ReplaceQuoteMachine.ZERO;
+            for (int i = 0; i < patternString.length(); ++i) {
+                cur = cur.next(sb, patternString.charAt(i));
+            }
+            cur = cur.EOF(sb);
+            return sb.toString();
+        }
+
+        @Override
+        public AbstractBaseMatcher macher(ImmutableBytesWritable ptr, SortOrder sortOrder) {
+            Preconditions.checkNotNull(ptr);
+            Preconditions.checkNotNull(sortOrder);
+            byte[] matcherSourceBytes = Utils.immutableBytesWritableToBytes(ptr, sortOrder);
```

Comment:
I hope there's a pattern.matcher(byte[] buf, int offset) method we can use instead below, as this will cause a copy of the underlying byte[]. 

If this is necessary, can you use ByteUtil.copyKeyBytesIfNecessary() as it prevents a copy unless necessary. I think this logic may be able to be pulled out of here and into LikeExpression.evaluate as it should be the same in both cases.